src/dune_rules/lock_dir.ml

```diff
@@ -250,15 +240,77 @@ let of_dev_tool_if_lock_dir_exists dev_tool =
   else Memo.return None
 ;;
 
+let lock_dirs_of_workspace (workspace : Workspace.t) =
+  let module Set = Path.Source.Set in
+  let+ lock_dirs_from_ctx =
+    Memo.List.map workspace.contexts ~f:(function
+      | Opam _ | Default { lock_dir = None; _ } -> Memo.return None
+      | Default { lock_dir = Some selection; _ } ->
+        let+ path = select_lock_dir selection in
+        Some path)
+    >>| List.filter_opt
+  in
+  match lock_dirs_from_ctx, workspace.lock_dirs with
+  | [], [] -> Set.singleton default_source_path
+  | lock_dirs_from_ctx, lock_dirs_from_toplevel ->
+    let lock_paths_from_toplevel =
+      List.map lock_dirs_from_toplevel ~f:(fun (lock_dir : Workspace.Lock_dir.t) ->
+        lock_dir.path)
+    in
+    Set.union (Set.of_list lock_paths_from_toplevel) (Set.of_list lock_dirs_from_ctx)
+;;
+
+let enabled () =
+  match !Clflags.ignore_lock_dir with
+  | true -> Memo.return false
+  | false ->
+    let* workspace = Workspace.workspace () in
+    (match workspace.config.pkg_enabled with
+     | Set (_, `Enabled) -> Memo.return true
+     | Set (_, `Disabled) -> Memo.return false
+     | Unset ->
+       let* lock_dirs = lock_dirs_of_workspace workspace in
+       Path.Source.Set.to_list lock_dirs
+       |> Memo.List.exists ~f:(fun lock_dir ->
+         Fs_memo.dir_exists (Path.Outside_build_dir.In_source_dir lock_dir)))
+;;
+
+let get_path_if_source_exists ctx_name =
+  let* workspace = Workspace.workspace () in
+  let ctx =
+    List.find_map workspace.contexts ~f:(fun ctx ->
+      match Context_name.equal (Workspace.Context.name ctx) ctx_name with
+      | false -> None
+      | true -> Some ctx)
+  in
+  let* lock_dir_paths =
+    match ctx with
+    | None | Some (Default { lock_dir = None; _ }) ->
+      Memo.return (Some (default_source_path, default_path))
+    | Some (Default { lock_dir = Some lock_dir_selection; _ }) ->
+      let+ source_lock_dir = select_lock_dir lock_dir_selection in
+      Some (source_lock_dir, lock_dir_of_source source_lock_dir)
+    | Some (Opam _) -> Memo.return None
```

Comment:
This looks like duplicated code from above (I haven't checked in detail, but that's kind of the point that it is rather dense logic that we don't want to duplicate). Probably worth factoring out.